src/utilities/include/par.h

```diff
@@ -49,47 +53,251 @@ inline constexpr ExecutionPolicy autoPolicy(size_t size) {
+#endif
+  std::exclusive_scan(first, last, d_first, init, f);
+}
+
+// TODO: use STL variant when parallelization is not enabled?
+
+template <typename InputIter, typename OutputIter, typename F>
+void transform(ExecutionPolicy policy, InputIter first, InputIter last,
+               OutputIter d_first, F f) {
+  for_each_n(policy, countAt(0_z), std::distance(first, last),
+             [&](size_t i) { d_first[i] = f(first[i]); });
+}
+
+template <typename InputIter, typename OutputIter>
+void copy(ExecutionPolicy policy, InputIter first, InputIter last,
+          OutputIter d_first) {
+  for_each_n(policy, countAt(0_z), std::distance(first, last),
+             [&](size_t i) { d_first[i] = first[i]; });
+}
+
+template <typename InputIter, typename OutputIter>
+void copy_n(ExecutionPolicy policy, InputIter first, size_t n,
+            OutputIter d_first) {
+  for_each_n(policy, countAt(0_z), n, [&](size_t i) { d_first[i] = first[i]; });
+}
+
+template <typename OutputIter, typename T>
+void fill(ExecutionPolicy policy, OutputIter first, OutputIter last, T value) {
+  for_each_n(policy, countAt(0_z), std::distance(first, last),
+             [&](size_t i) { first[i] = value; });
+}
+
+template <typename InputIter, typename P>
+size_t count_if(ExecutionPolicy policy, InputIter first, InputIter last,
+                P pred) {
+  return reduce(policy, TransformIterator(first, pred),
+                TransformIterator(last, pred), 0_z, std::plus<size_t>());
+}
+
+template <typename InputIter, typename P>
+bool all_of(ExecutionPolicy policy, InputIter first, InputIter last, P pred) {
+  // can probably optimize a bit for short-circuiting
+  return reduce(policy, TransformIterator(first, pred),
+                TransformIterator(last, pred), true,
+                [](bool a, bool b) { return a && b; });
+}
+
+#if MANIFOLD_PAR == 'T'
+template <typename InputIter, typename OutputIter, typename P>
+struct CopyIfScanBody {
+  size_t sum;
+  P &pred;
+  InputIter input;
+  OutputIter output;
+
+  CopyIfScanBody(P &pred, InputIter input, OutputIter output)
+      : sum(0), pred(pred), input(input), output(output) {}
+  CopyIfScanBody(CopyIfScanBody &b, tbb::split)
+      : sum(0), pred(b.pred), input(b.input), output(b.output) {}
+  template <typename Tag>
+  void operator()(const tbb::blocked_range<size_t> &r, Tag) {
+    size_t temp = sum;
+    for (size_t i = r.begin(); i < r.end(); ++i) {
+      if (pred(input[i])) {
+        temp += 1;
+        if (Tag::is_final_scan()) output[temp - 1] = input[i];
+      }
+    }
+    sum = temp;
+  }
+  size_t get_sum() const { return sum; }
+  void reverse_join(CopyIfScanBody &a) { sum = a.sum + sum; }
+  void assign(CopyIfScanBody &b) { sum = b.sum; }
+};
+#endif
+
+// note that you should not have alias between input and output...
+// in general it is impossible to check if there is any alias, as the input
+// iterator can be computed on-the-fly depending on the output position and may
+// not have a pointer
```

Comment:
Might be a good idea to copy the Thrust requirements documentation into each of these wrappers - I know I find myself looking those up constantly when I use these, and it would be more convenient to have them here anyway.

Reply:
Sure.